File: android/sample-ui/src/main/kotlin/com/usesmileid/sampleapps/ui/components/UseSmileIDSampleScanSheet.kt
package com.usesmileid.sampleapps.ui.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.WindowInsets
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.navigationBars
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.windowInsetsPadding
import androidx.compose.foundation.selection.selectable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.minimumInteractiveComponentSize
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.sp
import com.smileid.designsystem.SmileDimens
import com.usesmileid.sampleapps.ui.UseSmileIDSampleTestIds
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleEnvironment
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleSimulatedBindings
import com.usesmileid.sampleapps.ui.model.UseSmileIDSampleSimulatedSpan
import com.usesmileid.sampleapps.ui.theme.UseSmileIDSampleTheme

/** What the sheet renders, so the screen owns the entry state and the sheet stays stateless. */
@Immutable
data class UseSmileIDSampleScanSheetState(
    val token: String = "",
    /** Why the entered token is not a session — shown under the field, never the token itself. */
    val rejection: String? = null,
    val span: UseSmileIDSampleSimulatedSpan = UseSmileIDSampleSimulatedSpan.FifteenMinutes,
    /** Which host the minted token's `api_url` names, which is the whole of how a run picks its environment. */
    val environment: UseSmileIDSampleEnvironment = UseSmileIDSampleEnvironment.Sandbox,
    val bindings: UseSmileIDSampleSimulatedBindings = UseSmileIDSampleSimulatedBindings(),
    /** The mint controls start closed so the viewfinder keeps its height. */
    val expanded: Boolean = false,
)

/**
 * The sheet under the scanner: manual entry, and a simulated scan that mints its own fixture token.
 * Simulate is a product feature, not scaffolding — it is how a flow reaches the session states with
 * no QR source, and what it mints is chosen here rather than hard-coded.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun UseSmileIDSampleScanSheet(
    state: UseSmileIDSampleScanSheetState,
    onTokenChange: (String) -> Unit,
    onPaste: () -> Unit,
    onLink: () -> Unit,
    onExpandToggle: () -> Unit,
    onSpanSelect: (UseSmileIDSampleSimulatedSpan) -> Unit,
    onEnvironmentSelect: (UseSmileIDSampleEnvironment) -> Unit,
    onBindingsChange: (UseSmileIDSampleSimulatedBindings) -> Unit,
    onSimulate: () -> Unit,
    modifier: Modifier = Modifier,
) {
    val colors = UseSmileIDSampleTheme.colors
    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(topStart = SmileDimens.radiusSheet, topEnd = SmileDimens.radiusSheet),
        color = colors.surface,
    ) {
        Column(
            modifier = Modifier
                .fillMaxWidth()
                .windowInsetsPadding(WindowInsets.navigationBars)
                .padding(SmileDimens.spacingMd),
            verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingSm),
        ) {
            UseSmileIDSampleTextInput(
                value = state.token,
                onValueChange = onTokenChange,
                placeholder = "Or enter token manually",
                isError = state.rejection != null,
                errorMessage = state.rejection,
                // The token is a bearer credential and 900 characters long: nobody proofreads it, and
                // masked it stays out of screenshots and out of a failed run's hierarchy dump.
                masked = true,
                testId = UseSmileIDSampleTestIds.TOKEN_MANUAL_ENTRY,
                leading = { tint ->
                    Box(modifier = Modifier.size(SmileDimens.sizeIconMd), contentAlignment = Alignment.Center) {
                        ScanMarkGlyph(tint = tint)
                    }
                },
                trailing = {
                    Text(
                        text = "Paste",
                        style = UseSmileIDSampleTheme.type.linkFont.copy(
                            fontSize = SHEET_ACTION_SIZE,
                            fontWeight = FontWeight.Bold,
                        ),
                        color = colors.primary,
                        softWrap = false,
                        modifier = Modifier
                            .testTag(UseSmileIDSampleTestIds.TOKEN_PASTE)
                            .clickable(role = Role.Button, onClick = onPaste)
                            .minimumInteractiveComponentSize()
                            .padding(horizontal = SmileDimens.spacingXs),
                    )
                },
            )
            // Only once there is something to link, so the default sheet keeps the design's two rows.
            if (state.token.isNotBlank()) {
                UseSmileIDSampleButton(text = "Link token", onClick = onLink)
            }
            // Collapsed by default, and that is the point: this is a scanner, and the mint controls are a
            // probe affordance. Expanded they took enough height to leave the viewfinder a letterbox.
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable(role = Role.Button) { onExpandToggle() }
                    .padding(vertical = SmileDimens.spacingXxs),
                horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                UseSmileIDSampleSectionLabel(text = "SIMULATED SCAN", modifier = Modifier.weight(1f))
                Box(modifier = Modifier.size(SmileDimens.sizeIconMd), contentAlignment = Alignment.Center) {
                    if (state.expanded) {
                        ChevronDownGlyph(tint = colors.textMuted)
                    } else {
                        ChevronRightGlyph(tint = colors.textMuted)
                    }
                }
            }
            if (state.expanded) {
            FlowRow(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
                verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
            ) {
                UseSmileIDSampleSimulatedSpan.entries.forEach { span ->
                    ScanSheetChip(
                        label = span.label,
                        selected = state.span == span,
                        role = Role.RadioButton,
                        onClick = { onSpanSelect(span) },
                    )
                }
            }
            // Minting is where a run picks an environment, because there is no app-side control left.
            FlowRow(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
                verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
            ) {
                UseSmileIDSampleEnvironment.entries.forEach { environment ->
                    ScanSheetChip(
                        label = environment.label,
                        selected = state.environment == environment,
                        role = Role.RadioButton,
                        onClick = { onEnvironmentSelect(environment) },
                    )
                }
            }
            FlowRow(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
                verticalArrangement = Arrangement.spacedBy(SmileDimens.spacingXs),
            ) {
                ScanSheetChip(
                    label = "Binds consent",
                    selected = state.bindings.consent,
                    role = Role.Checkbox,
                    onClick = { onBindingsChange(state.bindings.copy(consent = !state.bindings.consent)) },
                )
                ScanSheetChip(
                    label = "Binds details",
                    selected = state.bindings.userDetails,
                    role = Role.Checkbox,
                    onClick = { onBindingsChange(state.bindings.copy(userDetails = !state.bindings.userDetails)) },
                )
            }
            }
            UseSmileIDSampleButton(
                text = "Simulate a successful scan",
                onClick = onSimulate,
                testId = UseSmileIDSampleTestIds.TOKEN_SIMULATE,
            )
        }
    }
}

/** The filter chip's shape without its count, because what a simulated scan mints has no count. */
@Composable
private fun ScanSheetChip(label: String, selected: Boolean, role: Role, onClick: () -> Unit) {
    val colors = UseSmileIDSampleTheme.colors
    Surface(
        modifier = Modifier
            .minimumInteractiveComponentSize()
            .selectable(selected = selected, role = role, onClick = onClick),
        shape = RoundedCornerShape(SmileDimens.radiusChip),
        color = if (selected) colors.primary else colors.filterChip.background,
        border = if (selected) null else BorderStroke(SmileDimens.borderWidthHairline, colors.filterChip.border),
    ) {
        Text(
            text = label,
            style = UseSmileIDSampleTheme.type.filterChipFont.copy(
                fontSize = SHEET_ACTION_SIZE,
                fontWeight = FontWeight.Bold,
            ),
            color = if (selected) colors.onPrimary else colors.filterChip.label,
            modifier = Modifier
                .defaultMinSize(minHeight = SmileDimens.space32)
                .padding(horizontal = SmileDimens.spacingSm, vertical = SmileDimens.spacingXs),
        )
    }
}

private val SHEET_ACTION_SIZE = 13.sp

File: ios/SampleUI/Sources/SampleUI/Components/UseSmileIDSampleScanSheet.swift
import SwiftUI

/// What the sheet edits. Lifted to the app state by the shell, because one tab is mounted at a time
/// and anything typed has to survive the screen that took it (R6).
public struct UseSmileIDSampleScanSheetState: Equatable, Sendable {
  public var token: String
  /// Why the entered token is not a session — shown under the field, never the token itself.
  public var rejection: String?
  public var span: UseSmileIDSampleSimulatedSpan
  /// Which host the minted token's `api_url` will name.
  public var environment: UseSmileIDSampleEnvironment
  public var bindings: UseSmileIDSampleSimulatedBindings
  /// The mint controls start closed so the viewfinder keeps its height.
  public var expanded: Bool

  public init(
    token: String = "",
    rejection: String? = nil,
    span: UseSmileIDSampleSimulatedSpan = .fifteenMinutes,
    environment: UseSmileIDSampleEnvironment = .sandbox,
    bindings: UseSmileIDSampleSimulatedBindings = UseSmileIDSampleSimulatedBindings(),
    expanded: Bool = false
  ) {
    self.token = token
    self.rejection = rejection
    self.span = span
    self.environment = environment
    self.bindings = bindings
    self.expanded = expanded
  }
}

/// The sheet under the scanner: manual entry, and a simulated scan that mints its own fixture token.
/// Simulate is a product feature, not scaffolding — it is how a flow reaches the session states with
/// no QR source, and what it mints is chosen here rather than hard-coded.
public struct UseSmileIDSampleScanSheet: View {
  @Binding private var state: UseSmileIDSampleScanSheetState
  private let onPaste: () -> Void
  private let onLink: () -> Void
  private let onSimulate: () -> Void

  @ScaledMetric(relativeTo: .body) private var actionHeight: CGFloat = SmileSpacing.sizeControlMd
  @Environment(\.useSmileIDSampleColors) private var colors
  @Environment(\.sizeCategory) private var sizeCategory

  public init(
    state: Binding<UseSmileIDSampleScanSheetState>,
    onPaste: @escaping () -> Void,
    onLink: @escaping () -> Void,
    onSimulate: @escaping () -> Void
  ) {
    _state = state
    self.onPaste = onPaste
    self.onLink = onLink
    self.onSimulate = onSimulate
  }

  public var body: some View {
    VStack(alignment: .leading, spacing: SmileSpacing.spacingSm) {
      UseSmileIDSampleTextInput(
        value: Binding(
          get: { state.token },
          set: { token in
            state.token = token
            state.rejection = nil
          }
        ),
        placeholder: "Or enter token manually",
        isError: state.rejection != nil,
        errorMessage: state.rejection,
        // The token is a bearer credential and 900 characters long: nobody proofreads it, and
        // masked it stays out of screenshots and out of a failed run's hierarchy dump.
        masked: true,
        testId: UseSmileIDSampleTestIds.tokenManualEntry
      ) {
        UseSmileIDSampleIcon(SmileIcons.tokenScan, tint: colors.input.placeholder, size: SmileSpacing.sizeIconMd)
      } trailing: {
        // Beside the field until type grows; then below it, or the placeholder breaks mid-word.
        if !sizeCategory.isAccessibilityCategory {
          pasteAction
        }
      }
      if sizeCategory.isAccessibilityCategory {
        pasteAction.frame(maxWidth: .infinity, alignment: .trailing)
      }
      // Only once there is something to link, so the default sheet keeps the design's two rows.
      if !state.token.isBlank {
        UseSmileIDSampleButton(text: "Link token", action: onLink)
      }
      // Collapsed by default, and that is the point: this is a scanner, and the mint controls are a
      // probe affordance. Expanded they took enough height to leave the viewfinder a letterbox.
      Button {
        state.expanded.toggle()
      } label: {
        HStack(spacing: SmileSpacing.spacingXs) {
          UseSmileIDSampleSectionLabel("SIMULATED SCAN")
            .frame(maxWidth: .infinity, alignment: .leading)
          UseSmileIDSampleIcon(
            state.expanded ? SmileIcons.chevronDown : SmileIcons.chevron,
            tint: colors.textMuted,
            size: SmileSpacing.sizeIconMd
          )
        }
        .padding(.vertical, SmileSpacing.spacingXxs)
        .contentShape(Rectangle())
      }
      .buttonStyle(.plain)
      if state.expanded {
        mintControls
      }
      UseSmileIDSampleButton(
        text: "Simulate a successful scan",
        testId: UseSmileIDSampleTestIds.tokenSimulate,
        action: onSimulate
      )
    }
    .padding(SmileSpacing.spacingMd)
    .frame(maxWidth: .infinity)
    // Its ideal height, not a share of what is left: compressed, its texts truncate at large type.
    .fixedSize(horizontal: false, vertical: true)
    .background(
      UseSmileIDSampleTopRoundedShape(radius: UseSmileIDSampleShapes.sheet)
        .fill(colors.surface)
        .ignoresSafeArea(edges: .bottom)
    )
  }

  private var pasteAction: some View {
    Button(action: onPaste) {
      UseSmileIDSampleText("Paste", style: UseSmileIDSampleTheme.type.linkFont.with(size: 13, weight: 700))
        .foregroundColor(colors.primary)
        .lineLimit(1)
        .padding(.horizontal, SmileSpacing.spacingXs)
        .frame(minHeight: actionHeight)
        .contentShape(Rectangle())
    }
    .buttonStyle(.plain)
    .useSmileIDSampleTestId(UseSmileIDSampleTestIds.tokenPaste)
  }

  @ViewBuilder
  private var mintControls: some View {
    chipRow {
      ForEach(UseSmileIDSampleSimulatedSpan.allCases, id: \.self) { span in
        ScanSheetChip(label: span.label, selected: state.span == span) { state.span = span }
      }
    }
    // Minting is where a run picks an environment, because there is no app-side control left.
    chipRow {
      ForEach(UseSmileIDSampleEnvironment.allCases, id: \.self) { environment in
        ScanSheetChip(
          label: environment.label,
          selected: state.environment == environment,
          testId: UseSmileIDSampleTestIds.tokenEnvironment(environment.id)
        ) { state.environment = environment }
      }
    }
    chipRow {
      ScanSheetChip(label: "Binds consent", selected: state.bindings.consent) { state.bindings.consent.toggle() }
      ScanSheetChip(label: "Binds details", selected: state.bindings.userDetails) { state.bindings.userDetails.toggle() }
    }
  }

  /// Stacks once type grows; SwiftUI has no FlowRow on this floor, so the switch is explicit.
  @ViewBuilder
  private func chipRow(@ViewBuilder _ chips: () -> some View) -> some View {
    if sizeCategory.isAccessibilityCategory {
      VStack(alignment: .leading, spacing: SmileSpacing.spacingXs) { chips() }
    } else {
      HStack(spacing: SmileSpacing.spacingXs) { chips() }
    }
  }
}

/// The filter chip's shape without its count, because what a simulated scan mints has no count.
private struct ScanSheetChip: View {
  let label: String
  let selected: Bool
  var testId: String?
  let onTap: () -> Void

  @ScaledMetric(relativeTo: .body) private var minHeight: CGFloat = SmileSpacing.space32
  @ScaledMetric(relativeTo: .body) private var touchHeight: CGFloat = SmileSpacing.sizeControlMd
  @Environment(\.useSmileIDSampleColors) private var colors

  var body: some View {
    Button(action: onTap) {
      UseSmileIDSampleText(label, style: UseSmileIDSampleTheme.type.filterChipFont.with(size: 13, weight: 700))
        .foregroundColor(selected ? colors.onPrimary : colors.filterChip.label)
        .padding(.horizontal, SmileSpacing.spacingSm)
        .padding(.vertical, SmileSpacing.spacingXs)
        .frame(minHeight: minHeight)
        .background(
          RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.chip, style: .continuous)
            .fill(selected ? colors.primary : colors.filterChip.background)
        )
        .overlay(
          RoundedRectangle(cornerRadius: UseSmileIDSampleShapes.chip, style: .continuous)
            .strokeBorder(selected ? .clear : colors.filterChip.border, lineWidth: SmileSpacing.borderWidthHairline)
        )
        // The platform's touch target around the design's 32pt chip.
        .frame(minHeight: touchHeight)
        .contentShape(Rectangle())
    }
    .buttonStyle(.plain)
    .accessibilityAddTraits(selected ? [.isButton, .isSelected] : .isButton)
    .useSmileIDSampleTestId(testId)
  }
}

/// A surface rounded on its top corners only, which the iOS 15 floor has no shape for.
struct UseSmileIDSampleTopRoundedShape: Shape {
  let radius: CGFloat

  func path(in rect: CGRect) -> Path {
    var path = Path()
    path.move(to: CGPoint(x: rect.minX, y: rect.maxY))
    path.addLine(to: CGPoint(x: rect.minX, y: rect.minY + radius))
    path.addArc(
      center: CGPoint(x: rect.minX + radius, y: rect.minY + radius),
      radius: radius,
      startAngle: .degrees(180),
      endAngle: .degrees(270),
      clockwise: false
    )
    path.addLine(to: CGPoint(x: rect.maxX - radius, y: rect.minY))
    path.addArc(
      center: CGPoint(x: rect.maxX - radius, y: rect.minY + radius),
      radius: radius,
      startAngle: .degrees(270),
      endAngle: .degrees(0),
      clockwise: false
    )
    path.addLine(to: CGPoint(x: rect.maxX, y: rect.maxY))
    path.closeSubpath()
    return path
  }
}
